modules/siddhi-extensions/kf/src/main/java/org/wso2/siddhi/extension/kf/Kalman2DFilterWithMatrices.java

```diff
@@ -0,0 +1,267 @@
+
+        //Xk = (A * Xk-1)
+        prevMeasuredMatrixX = transitionMatrixA.multiply(prevMeasuredMatrixX);
+        prevMeasuredMatrixY = transitionMatrixA.multiply(prevMeasuredMatrixY);
+
+        //Pk = (A * P * AT) + Q
+        varianceMatrixP = (transitionMatrixA.multiply(varianceMatrixP)).multiply(transitionMatrixA.transpose());
+
+        //S = (H * P * HT) + R
+        RealMatrix S = ((measurementMatrixH.multiply(varianceMatrixP)).multiply(measurementMatrixH.transpose())).add(Rmatrix);
+        RealMatrix S_1 = new LUDecomposition(S).getSolver().getInverse();
+
+        //P * HT * S-1
+        RealMatrix kalmanGainMatrix = (varianceMatrixP.multiply(measurementMatrixH.transpose())).multiply(S_1);
+
+        //Xk = Xk + kalmanGainMatrix (Zk - HkXk )
+        prevMeasuredMatrixX = prevMeasuredMatrixX.add(kalmanGainMatrix.multiply(
+                (measuredMatrixX.subtract(measurementMatrixH.multiply(prevMeasuredMatrixX)))));
+
+        prevMeasuredMatrixY = prevMeasuredMatrixY.add(kalmanGainMatrix.multiply(
+                (measuredMatrixY.subtract(measurementMatrixH.multiply(prevMeasuredMatrixY)))));
+
+        //Pk = Pk - K.Hk.Pk
+        varianceMatrixP = varianceMatrixP.subtract(
+                (kalmanGainMatrix.multiply(measurementMatrixH)).multiply(varianceMatrixP));
+
+        prevTimestamp = timestamp;
+        kalmonFilter.setPrevMeasuredMatrixX(prevMeasuredMatrixX);
+        kalmonFilter.setPrevMeasuredMatrixY(prevMeasuredMatrixY);
+        kalmonFilter.setVarianceMatrixP(varianceMatrixP);
+        kalmonFilter.setPrevTimestamp(prevTimestamp);
+        kalmonFilter.setMeasurementMatrixH(measurementMatrixH);
+        kalmonFilterHashMap.get(objectId).put(level, kalmonFilter);
+        return new Object[]{prevMeasuredMatrixX.getRow(0)[0], prevMeasuredMatrixY.getRow(0)[0]};
+    }
+
+    @Override
+    protected Object[] process(Object data) {
+        return new Object[0];
+    }
+
+    @Override
+    protected List<Attribute> init(AbstractDefinition inputDefinition,
+                                   ExpressionExecutor[] attributeExpressionExecutors,
+                                   ExecutionPlanContext executionPlanContext) {
+        kalmonFilterHashMap = new ConcurrentHashMap<String, ConcurrentHashMap<Integer, KalmonFilter>>();
+        if (attributeExpressionExecutors.length != 7) {
+            throw new ExecutionPlanValidationException("Invalid no of arguments passed to " +
+                                                       "kf:kalman2DFilter() function, " +
+                                                       "requires 7, but found " + attributeExpressionExecutors.length);
+        }
+        ArrayList<Attribute> attributes = new ArrayList<Attribute>(2);
+        attributes.add(new Attribute("kalmonLatitude", Attribute.Type.DOUBLE));
+        attributes.add(new Attribute("kalmonLongitude", Attribute.Type.DOUBLE));
+        return attributes;
+    }
+
+    private class KalmonFilter {
```

Comment:
@lasanthafdo since it contains all information relates to the calculation of the Kalman filter, shall we make it KFRelatedValueHolder? wdyt? or CompositeMeasurement would be good too.